HHKB modにnicola追加

aaa70 opened the pull request:
qmk_firmwareへのnicola実装と公開ありがとうございます。
crkbdのnicolaをベースにHHKB hasuコントローラに対応してみました。
このようにPRを送っていいのかわかりませんが、お送りしたいと思います。

nicola2, nicola3を参考にしつつ下記の変更をしてあります。

* レイヤーキーを三つ固定から可変に
* NLSHFT, NRSHFTは使用せず、既存のキーを使用し単独押下に対応
* modifierキーの同時押し対応
* modifierキーリリース時に押下されているキーがある場合の連打防止

私のhasuコントローラはbluetoothに対応したものではないので、bluetoothで動作するかは未検証です。

また本家のqmk_firmwareへの個人のキーマップ登録の流れなどは全く知らないのですが本家に取り込んでもらう予定などはないのでしょうか。
よろしくお願い致します。

eswai commented:
PRありがとうございます。
本家へのマージはいつかやりたいですが、私も流れを理解できていません。

aaa70 commented:
マージありがとうございます。
あまりメリットはないかもしれませんが、本家でnicola, naginata実装がお目見えするのを楽しみにしております。

Bamboo-inside commented:
HHKB hasu controllerのnicola公開いただきありがとうございます。
cloneしてコンパイルまでこぎつけ，してみたのですが，以下のエラーが出ました。

Compiling: keyboards/hhkb/keymaps/nicola/nicola.c                                                  keyboards/hhkb/keymaps/nicola/nicola.c: In function 'process_nicola':
keyboards/hhkb/keymaps/nicola/nicola.c:103:35: error: 'sizeof' on array function parameter 'modkeys' will return size of 'uint16_t *' {aka 'unsigned int *'} [-Werror=sizeof-array-argument]
  103 |   static int modkeys_size = sizeof(modkeys) / sizeof(modkeys[0]);
      |                                   ^
keyboards/hhkb/keymaps/nicola/nicola.c:102:128: note: declared here
  102 | bool process_nicola(uint16_t keycode, keyrecord_t *record, uint8_t ncl_layer, uint16_t lshiftkey, uint16_t rshiftkey, uint16_t modkeys[]) {
      |                                                                                                                       ~~~~~~~~~^~~~~~~~~
cc1: all warnings being treated as errors
 [ERRORS]
 |
 |
 |
make[1]: *** [.build/obj_hhkb_nicola/nicola.o] Error 1
make: *** [hhkb:nicola] Error 1
Make finished with errors

コード詳しくないのですみません。ただ，
sizeof(modkeys)をsizeof(*modkeys)
とするととりあえずコンパイルは通るのですが，これはaaa70様の意図する内容ですか？
このエラーの正しい解決を教えていただけると幸いです。

（コメントをどこにするのがよいのかよくわからずここに投稿しています）

aaa70 commented:
コンパイルした環境をお聞きしてよろしいでしょうか。またこちらはcloneしてそのまま"make hhkb:nicola"をしてエラーになりましたか。

modkeys自体はニコラ配列使用中に別のレイヤーを使用するのに必要な場合があるかと思います。
eswaiさんのcrkbd/nicolaオリジナルでは3つのレイヤーを固定で指定されていたのですがレイヤーの数に応じて動的に変更できるように私が変更しました。

keymap.cにて"static uint16_t mkeys[] = {FN};"のまま使用されている場合"sizeof(*modkeys)"でも1が返るだけなので問題なく動作はすると思います。

別のレイヤーの処理が不要ということであれば下記のようにして処理自体なくしてしまってもいいかもしれません。
```diff
modified   keyboards/hhkb/keymaps/nicola/keymap.c
@@ -12,8 +12,6 @@ enum custom_keycodes {
   FN
 };
 
-static uint16_t mkeys[] = {FN};
-
 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     [_QWERTY] = LAYOUT_JP(
         KC_ESC, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL, KC_JYEN, KC_BSPC,
@@ -84,6 +82,6 @@ bool process_record_user(uint16_t keycode, keyrecord_t *record) {
   }
 
   // 親指シフト
-  return process_nicola(keycode, record, _NICOLA, KC_SPC, KC_HENK, mkeys);
+  return process_nicola(keycode, record, _NICOLA, KC_SPC, KC_HENK);
   // 親指シフト
 }
modified   keyboards/hhkb/keymaps/nicola/nicola.c
@@ -99,9 +99,7 @@ void ncl_clear(void) {
   ncl_rshift = false;
 }
 
-bool process_nicola(uint16_t keycode, keyrecord_t *record, uint8_t ncl_layer, uint16_t lshiftkey, uint16_t rshiftkey, uint16_t modkeys[]) {
-  static int modkeys_size = sizeof(modkeys) / sizeof(modkeys[0]);
-
+bool process_nicola(uint16_t keycode, keyrecord_t *record, uint8_t ncl_layer, uint16_t lshiftkey, uint16_t rshiftkey) {
   switch (keycode) {
   case KC_LCTRL:
   case KC_LSHIFT:
@@ -119,16 +117,6 @@ bool process_nicola(uint16_t keycode, keyrecord_t *record, uint8_t ncl_layer, ui
     }
     break;
   }
-  for (int i = 0; i < modkeys_size; i++) {
-    if (keycode == modkeys[i]) {
-      if (record->event.pressed) {
-        ncl_modcount++;
-      } else {
-        ncl_modcount--;
-        if (ncl_modcount == 0) ncl_modrelease = true;
-      }
-    }
-  }
 
   if (ncl_modcount > 0) {
     return true;
modified   keyboards/hhkb/keymaps/nicola/nicola.h
@@ -16,4 +16,4 @@
 
 void ncl_type(void);
 void ncl_clear(void);
-bool process_nicola(uint16_t, keyrecord_t *, uint8_t, uint16_t, uint16_t, uint16_t[]);
+bool process_nicola(uint16_t, keyrecord_t *, uint8_t, uint16_t, uint16_t);
```

aaa70 commented:
すみません、私の知識不足と環境が古いのが原因のようでした。先程、新たに #3 のPRを出しました。

aaa70 commented:
@eswai さん、マージありがとうございました。お手数おかけいたしました。
@Bamboo-inside さん、マージされたもので改めて試していただいてよろしいでしょうか。

Bamboo-inside commented:
@aaa70 さん，ありがとうございます。
cloneそのままでコンパイル通りました。
QMK Firmware 0.9.5でもコンパイルできました。
